[Joy] step-1/step2 구현

he2joojo opened the pull request:
안녕하세요 kyu! 😃 5주동안 리뷰를 받게 된 조이입니다! 첫 리뷰라 혼자 작업할 때 처럼 커밋을 너무 작게 나눴습니다. 다음 리뷰때부터는 커밋 단위를 개선하겠습니다! 

### 1️⃣ **작업 내용**
- [x] 타임리프를 이용해 중복된 html코드를 분리했습니다.
- [x] 회원가입 기능을 구현했습니다. 중복된 아이디로 가입시, 저장하지 않고 다시 회원가입 창으로 보냈습니다.
- [x] 회원 목록 조회 기능을 구현했습니다. 목록에서 회원 이름을 클릭하면 개인 프로필로 이동하게 했습니다. 
- [x] 글쓰기 기능을 구현했습니다.
- [x] 작성한 글 목록을 index화면에서 볼 수 있게 구현했습니다.
- [x] 게시글 상세 내용을 볼 수 있게 했습니다.
____
### 2️⃣ **학습 키워드**
 * Thymeleaf 기본적인 사용법
 * 스프링에서 빈을 수동으로 지정하는 방법과 자동으로 등록하는 방법
 * model.addAttribute 사용법
 * @PathVariable가 무엇인가?
____
### 3️⃣ **고민과 해결**
- [x] 상대경로와 절대경로 문제로 시간을 많이 썼습니다. html 파일의 경로를 여러번 수정해 css와 js가 적용되지 않는 문제를 해결하고 url이 누적되는 문제를 해결했습니다.
- [x] 타임리프 적용문제 : td태그안에 th:text를 쓰고 연속해서 a태그 안에 th:href를 썼더니 a태그가 적용이 안되어 타임리프를 모두 a태그로 이동시켜 해결했습니다. ```<td th:text="${user.name}"><a th:href="@{/users/{userId}(userId=${user.userId})}">name</a></td>``` ➡️ ```<td ><a th:text="${user.name}" th:href="@{/users/{userId}(userId=${user.userId})}">name</a></td>```
- [ ] WebMvcConfigurer를 구현한 MvcConfig클래스를 꼭 사용해야 하는 이유를 아직 몰라 주석으로 처리해두었습니다.
- [ ] 글을 여러개 작성한 후 홈화면에 글 번호가 겹쳐 보이는 문제 : 타임리프의 문제인지 html 태그 문제인지 아직 파악하지 못했습니다.
____
🔅 처음 들어갔을 때의 홈 화면
<img width="1470" alt="스크린샷 2023-03-31 오후 5 38 46" src="https://user-images.githubusercontent.com/121915790/229075550-b1d0ba8a-2cbd-4e52-a032-d3e4e5116b2d.png">

🔅 정상 회원가입 후 회원 목록 화면
<img width="1470" alt="스크린샷 2023-03-31 오후 5 39 46" src="https://user-images.githubusercontent.com/121915790/229075637-0948b627-cc13-4f0f-9905-95c2b9ddec99.png">

🔅 여러명 회원가입 후 회원 목록 화면
<img width="1470" alt="스크린샷 2023-03-31 오후 5 40 45" src="https://user-images.githubusercontent.com/121915790/229076017-a062bc3c-53cc-4831-a37c-9984a9ea8546.png">

🔅 게시글 작성 완료 후 홈 화면 
<img width="1470" alt="스크린샷 2023-03-31 오후 5 41 17" src="https://user-images.githubusercontent.com/121915790/229075899-03e6f6aa-3ae8-44c7-bb91-59a5f93eb710.png">

🔅 게시글 상세 화면 (아직 댓글달기 기능은 구현하지 않았습니다)
<img width="1470" alt="스크린샷 2023-03-31 오후 5 41 24" src="https://user-images.githubusercontent.com/121915790/229076202-ac682baa-1b3b-44c0-af63-818f7e77b759.png">

🔅 개인 프로필 조회 화면
<img width="1470" alt="스크린샷 2023-03-31 오후 5 40 08" src="https://user-images.githubusercontent.com/121915790/229076287-b56a5e0a-0549-4c0d-ad00-af3b81e0b5d9.png">

⁉️ 글을 여러개 작성했을 때 홈 화면
<img width="1470" alt="스크린샷 2023-03-31 오후 6 23 11" src="https://user-images.githubusercontent.com/121915790/229081657-378ff2fc-4f2a-474d-bec0-de22e4257820.png">

Commit:
docs : 학습계획 초안 작성

Commit:
docs : 학습 계획 수정

Commit:
docs : 오늘 한 일 체크

kyupid commented:
안녕하세요 조이~ 처음 뵙겠습니다 이번 5주동안 리뷰 맡게된 큐라고 합니다
궁금한게 있는데 뒤에 xx분은 동영상 길이인가요~? 어떤거보고 학습하시는지 여쭤봐도 될까요?

he2joojo commented:
kyu 안녕하세요!! 
강의 동영상 길이 맞습니다~! 김영한님 강의 스프링 MVC1편 듣고있습니다!! :)

Commit:
docs : buid.gradle에 타임리프 추가

Commit:
feat : 패키지 생성, 기본 클래스 생성

Commit:
docs : html파일 a태그 경로에 .html 제거

Commit:
feat: 사용자 정보 저장하는 User클래스 생성

Commit:
feat : 사용자 입력 저장하는 UserForm클래스 생성

Commit:
feat : UserRepository 인터페이스 생성

user정보를 저장하고, 검색하는 기능을 가진 인터페이스를 만들었습니다.

Commit:
feat : UserRepository의 구현클래스 MemoryUserRepository 생성

private변수로 User를 저장하는 ConcurrentHashMap을 가진 클래스로, UserRepository를 구현한 클래스를 만들었습니다.

Commit:
feat : UserService 클래스 생성

private 변수로 userRepository를 가지고 있어서, 회원가입과 회원 조회 서비스를 제공하는 클래스 입니다.

Commit:
feat: HomeController 와 UserController 추가

html파일에 따라 작동하는 메서드를 가진 클래스를 추가했습니다.

Commit:
refactor : UserForm 변수명 변경

Commit:
refactor : Controller클래스의 메서드의 html경로 변경

Commit:
refactor : html파일의  a태그에 확장자 추가

.html확장자를 제거했더니 템플릿을 찾을 수 없다는 에러가 떠서 다시 확장자를 붙였습니다.

Commit:
feat : HomeController에 GetMapping메서드 추가

Commit:
feat : UserControlle 메서드의 반환경로 수정

Commit:
refactor : 변수명 변경

Commit:
refactor : MemoryUserRepository와 UserService 변수명 변경

Commit:
fix : html파일의 a태그에 확장가 추가

.html 확장자를 제거했더니 오류가 나서 다시 전부 .html확장자를 추가했다.

Commit:
fix : 중복 GetMapping 제거

HomeController와 UserController에 중복 @GetMapping이 있어서 에러가 발생해 UserController의 createForm()을 제거했다.

Commit:
fix : html파일의 css파일 경로 수정

디렉토리를 여러번 옮겨서 각 html파일에 css파일 경로가 잘못 기재되어 css가 적용되지 않았다. index.html은 css.~으로 변경하고 나머지 파일은 /css.~로 변경하였다.

Commit:
fix : create메서드의 @PostMapping 경로 수정

Commit:
fix : method를 GET에서 POST로 변경

Commit:
fix : html파일에 js 경로 변경

Commit:
buid : thymeleaf, devtools 의존성 라이브러리 추가

Commit:
build : application.properties에 Thymeleaf 설정 추가

Commit:
style : .html 확장자 제거

Commit:
refactor : userId 타입을 String로 변경

Commit:
refactor :UserForm의  userId 타입을 String로 변경

Commit:
refactor :index. html과 profile.html 안의 .html확장자 제거

pie2457 commented:
서로 어려웠던 부분에 대해 의논하고, 생각해 보지 못했던 점을 이야기 하며 아주 화이팅 있는 시간을 보냈습니다 !
항상 학습계획 꼼꼼히 쓰시는 모습, 커밋 세세하게 하시는 부분까지!  꼭 배워가겠습니다 멋쨍ㅇ이 ☀️

Commit:
chore : 상대경로를 절대경로로 변경

Commit:
feat : Thymeleaf를 적용해서 사용자 목록조회 기능 추가

Commit:
feat : 회원 프로필 조회 기능 추가

Commit:
refactor : html 중복 제거

Commit:
feat : 사용자 이름을 누르면 프로필이 보이게 추가

Commit:
feat : 글 작성 기능추가를 위한 클래스 생성

Commit:
feat : 글 저장하는 기능 구현

Commit:
feat : 작성 글 목록 보기 기능 구현

Commit:
feat : 게시글 보기 기능 구현

Commit:
feat : Article에 localDateTime 추가

Commit:
chore : 상대경로 절대경로로 변경

Commit:
feat : 동일 아이디로 회원가입시 가입되지 않는 기능 추가

Commit:
chore : 불필요한 주석 제거

kyupid commented on src/main/java/kr/codesqaud/cafe/config/MvcConfig.java:
```diff
@@ -0,0 +1,15 @@
+//
+//    @Override
+//    public void addViewControllers(ViewControllerRegistry registry){}
+//}
```
제 생각엔 개발할 때 뭐든지 "꼭" 사용해야하는 이유는 없을 것 같아요. 
아래 질문들을 생각해보시면 질문에 도움이 좀 될 것 같습니다.

- 스프링 부트와 스프링은 어떤 차이를 가지고 있을까요? 
- 스프링 부트에서 자동으로 구성해주는 것들은 무엇일까요?
- WebMvcConfigurer 를 구현하면 어떤 일을 할 수 있을까요?

kyupid replied:
참고로 위 코멘트는 조이가 남겨둔 고민에 대한 답변입니다.

```
 WebMvcConfigurer를 구현한 MvcConfig클래스를 꼭 사용해야 하는 이유를 아직 몰라 주석으로 처리해두었습니다.
```

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,48 @@
+
+    private final ArticleService articleService;
+
+    @Autowired
+    public ArticleController(ArticleService articleService) {
+        this.articleService = articleService;
+    }
```
생성자가 하나만 있고, 그 생성자의 파라미터들이 모두 빈으로 등록되어 있으면 `@Autowired` 어노테이션 없이도 스프링이 자동으로 생성자를 통한 의존성 주입을 해줍니다. 따라서 해당 코드에서 `@Autowired`를 생략해도 의존성 주입이 정상적으로 이루어져요.

명확한 의도를 표현하기 위해 `@Autowired`를 사용하는 것도 좋은 방법이지만 코멘트 남기겠습니다.   
혹시 `@Autowired` 를 붙이면 무슨 일이 일어나는지 잘 모르신다면 꼭 먼저 공부해보세요. 😊

he2joojo replied:
```@Autowired```가 생략 가능한 경우가 있다는 걸 잊고있었네요!! 수정했습니다 😃

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @PostMapping("qna/form")
+    public String write(ArticleForm form) {
+
+        Article article = new Article(form.getWriter(), form.getTitle(), form.getContents());
+        articleService.write(article);
+
```
작성하신 다른 코드를 보면 개행하는 데에도 나름의 의미를 생각하신 것 같습니다. 이 부분도 깔끔하게 붙입시다 ㅎㅎ

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,48 @@
+        return "index";
+    }
+
+    @GetMapping("articles/{title}")
+    public String showArticle(Model model, @PathVariable String title) {
+        model.addAttribute("article", articleService.findOne(title));
+        return "qna/show";
+    }
```
`@GetMapping("articles/{title}")` 이 조금 어색하게 느껴지네요. `title` 을 `@PathVariable` 로 받으신 이유가 궁금해요

he2joojo replied:
동일한 작성자가 글을 여러개 작성한 경우가 있다고 생각해서, title를 ```@PathVariable``` 로 받았습니다!! 이후에 같은 글 제목을 두개 이상 생성할 수 없게 로직을 추가할 생각이었는데, 다시 보니 제가 봐도 조금 어색한 것 같네요... ```@ PathVariable```로 뭘 받으면 좋을 지 더 생각해보겠습니다!

kyupid replied:
> 같은 글 제목을 두개 이상 생성할 수 없게

동일한 작성자가 글을 여러개 작성한다는 게 작성자가 두 번이상 요청하는 것을 방지하기 위함인가요?

he2joojo replied:
동일한 작성자가 글을 여러개 쓰는 것은 가능한데, 제목이 같은 글은 하나만 작성할 수 있습니다! 
> 동일한 작성자가 글을 여러개 작성한 경우
<img width="1470" alt="스크린샷 2023-04-03 오전 11 09 42" src="https://user-images.githubusercontent.com/121915790/229396166-cbd31d49-52f4-4013-8795-cb741014dd28.png">

>  이미 존재하는 글과 같은 제목으로 글을 작성시 에러 발생
<img width="1470" alt="스크린샷 2023-04-03 오전 11 11 13" src="https://user-images.githubusercontent.com/121915790/229396242-27c791e5-0508-46a3-bb07-5b6e37cf371b.png">

처음에 ```@GetMapping("articles/{userId}")```로 userId를 ```@PathVariable``` 로 받아서 작성 글을 보여주려고 했는데, 한 작성자가 글을 여러개 작성하면 url이 중복돼서 어떤 글을 보여줘야 할 지 모호해 진다는 문제가 생겼습니다. <br>

그래서 title 을 ```@PathVariable``` 로 받고, 대신 같은 제목을 두개 이상 생성할 수 없게해서 url이 중복되는 문제를 해결하려 했습니다!!

kyupid replied:
의도는 이해했습니다. Article 에 Id 필드를 하나 추가해주면 어떨까요?

he2joojo replied:
그런 방법이 있었네요!! title을 ```@PathVariable```로 받는게 계속 걸렸는데 감사합니다🥺

kyupid replied:
많이 고민하신 흔적이 보이네요. 지금처럼 계속 화이팅입니다💪

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+
+    @PostMapping("user/create")
+    public String create(UserForm form) {
+        User user = new User(form.getUserId(), form.getPassword(), form.getName(), form.getEmail());
```
지금은 괜찮지만 User 에 더 많은 필드가 있으면 생성자 파라미터가 굉장히 많고 헷갈릴거 같은데요. 다음엔 빌더 패턴도 한 번 고려해봅시다. 😊 

[참고 자료](https://johngrib.github.io/wiki/pattern/builder/#effective-java%EC%9D%98-%EB%B9%8C%EB%8D%94-%ED%8C%A8%ED%84%B4)

he2joojo replied:
좋은자료 감사합니다! 빌더패턴이 뭔지 이해했습니다!!😃

kyupid replied:
벌써 이해하셨다니 최곱니다👍

kyupid commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,45 @@
+    private void validateDuplicateUser(User user) {
+        userRepository.findById(user.getUserId())
+                .ifPresent(m -> {
+                    throw new IllegalStateException("이미 존재하는 아이디입니다");
```
자바에서 다른 IllegalStateException 포함 다른 예외들은 어떻게 활용되고 있을까요? 

아래 표도 참고해보시고 한 번 조사해보세요 😊
![2023-04-01_11-01-03](https://user-images.githubusercontent.com/59721293/229260550-b290f698-689d-4bd9-81ab-1a860e8e8532.jpg)

kyupid commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,45 @@
+        }
+    }
+
+    private void validateDuplicateUser(User user) {
+        userRepository.findById(user.getUserId())
```
아래 `findOne(String userId)` 와 비슷한 기능을 하는 것 같아요. 합쳐서 사용해볼 순 없을까요?

kyupid commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public String join(User user) {
+        try {
+            validateDuplicateUser(user);   // 중복아이디라면, user를 저장하지 않고 " "반환
+            userRepository.save(user);
+            return user.getUserId();
+        } catch (IllegalStateException e) {
+            return " ";
```
세 가지 의견이 있습니다.

1. `" "` 라는 표현을 사용하고 싶으셨다면 주석으로 의미를 전달하는 것도 좋지만 변수에 이름을 붙여서 좀 더 명확하게 표현해야 바람직할 것 같습니다. 그리고 불변이면 더 좋겠죠.  왜 그럴까요?
2. join 메소드에서 try-catch 문이 필요해보이진 않습니다. 동시에 `" "`와 같은 상수도 불필요해 보입니다. 개선해볼 수 없을까요?
3. IllegalStateException 은 try-catch 로 감싸지 않아도 되는 Exception 입니다. Checked Exception 과 Unchecked Exception 에 대해 공부해보세요.

he2joojo replied:
그룹리뷰에서 예외를 try-catch로 처리하지 말고 ```@Controlleradvice```로 다뤄보라는 피드백을 받아 공부중인데 아직은 너무 어려워  일단 UserService의 join메서드를 ```public boolean join(User user) {
       if (findOne(user.getUserId()).isPresent()) {
           return false;
       }
       userRepository.save(user);
       return true;
    }``` 로 임시 변경했습니다!! 좋은 피드백 감사합니다!:)

kyupid replied:
ControllerAdvice 어노테이션 배워서 사용하는 것도 좋구요. 개인적으론 지금처럼 날것으로 어떻게 해야 문제를 풀 수 있을지 고민하는게 더 성장에 도움 될거같아요. 잘하셨습니다 👏

kyupid commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+
+        String id = userService.join(user);
+
+        if (" ".equals(id)) {            // 중복아이디면, 회원가입 창으로
+            return "redirect:/user/form";
+        } else {
+            return "redirect:/users";    // 정상가입되었다면, 회원목록 창으로
```
해당 코드는 UserService 에 대한 리뷰를 참고해주세요.

kyupid commented on src/main/java/kr/codesqaud/cafe/dto/UserForm.java:
```diff
@@ -0,0 +1,41 @@
+
+    public void setEmail(String email) {
+        this.email = email;
+    }
```
이 클래스의 setter 들은 언제 사용하는 것인가요?

he2joojo replied:
setter들을 삭제하면 회원가입시 에러가 발생해서 모두 살려놨습니다! 왜 에러가 발생하는지는 스프링을 더 공부해봐야 알 것 같습니다🥹

kyupid replied:
타임리프에서 setter 를 사용하는거였군요 ㅠㅠ 제가 잘 몰랐습니다.

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,42 @@
+@Repository
+public class MemoryArticleRepository implements ArticleRepository {
+
+    private Map<String, Article> articleStore = new ConcurrentHashMap<>(); // 제목:Article 저장
```
HashMap 이 아니라 ConcurrentHashMap 을 사용하신 이유가 무엇인가요?

he2joojo replied:
웹사이트라서 동시성 문제가 생길 수 있다고 생각해서 ConcurrentHashMap을 사용했습니다 :)

kyupid commented on src/main/java/kr/codesqaud/cafe/repository/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public List<Article> findAll() {
+        return new ArrayList<>(articleStore.values());
```
그냥 articleStore 을 반환하면 안될까요? 꼭 새로운 List 로 반환해야하나요? 어떤 이유로 이렇게 작성하셨을까요?

he2joojo replied:
articleStore를 그대로 반환하면 MemoryArticleRepository와 ArticleService클래스 사이의 결합도가 높아질 것 같아서 새로운 List로 만들어 반환했습니다!

kyupid replied:
결합도가 높다는 게 어떤 의미인가요? 🤔  Repository 에 있는 필드를 Service 에서 사용하면 문제가 생기나요? 그렇게 치면 service 에서 repository 메소드를 사용하는 것도 결합도가 높아서 사용하면 안되지 않을까요?

kyupid reviewed (changes requested):
안녕하세요 Joy🕺 첫 PR 고생 많으셨습니다. 👍

리뷰에 던진 질문들이 시간을 들여야하는 것들이 많을 것 같아요. 하실 수 있는 만큼 공부해보시고 기억해두었다가 다음에라도 다시 꼭 보시길 권장합니다. 다 중요한 것들이에요.

그리고 이번 주말에 join() 메소드 만큼은 꼭 리팩토링 해보시길 바랍니다. 일요일까지 기다리고 있을게요 😊
